src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -81,58 +85,77 @@ internal abstract partial class DocumentProvider
-            // hierarchy if it is not.
-            if (RunningDocumentTable.TryGetCookieForInitializedDocument(documentKey.Moniker, out foundCookie))
-            {
-                object foundDocData = RunningDocumentTable.GetDocumentData(foundCookie);
-                openTextBuffer = TryGetTextBufferFromDocData(foundDocData);
-                if (openTextBuffer == null)
+            if (IsForeground())
+            {
+                // If we are on the foreground thread and this document is already open in the editor we want to associate the text buffer with it.
+                // However, determining if a document is already open is a bit complicated. With the introduction
+                // of the lazy tabs feature in Dev12, a document may be open (i.e. it has a tab in the shell) but not
+                // actually initialized (no data has been loaded for it because its contents have not actually been
+                // viewed or used). We only care about documents that are open AND initialized.
+                // That means we can't call IVsRunningDocumentTable::FindAndLockDocument to find the document; if the
+                // document is open but not initialized, the call will force initialization. This is bad for two
+                // reasons:
+                //   1.) It circumvents lazy tabs for any document that is part of a VB or C# project.
+                //   2.) Initialization may cause a whole host of other code to run synchronously, such as taggers.
+                // Instead, we check if the document is already initialized, and avoid asking for the doc data and
+                // hierarchy if it is not.
+                if (RunningDocumentTable.TryGetCookieForInitializedDocument(documentKey.Moniker, out foundCookie))
                 {
-                    // We're open but not open as a normal text buffer. This can happen if the
-                    // project system (say in ASP.NET cases) is telling us to add a file which
-                    // actually isn't a normal text file at all.
-                    return null;
-                }
+                    object foundDocData = RunningDocumentTable.GetDocumentData(foundCookie);
+                    openTextBuffer = TryGetTextBufferFromDocData(foundDocData);
+                    if (openTextBuffer == null)
+                    {
+                        // We're open but not open as a normal text buffer. This can happen if the
+                        // project system (say in ASP.NET cases) is telling us to add a file which
+                        // actually isn't a normal text file at all.
+                        return null;
+                    }
 
-                if (!canUseTextBuffer(openTextBuffer))
-                {
-                    return null;
+                    if (!canUseTextBuffer(openTextBuffer))
+                    {
+                        return null;
+                    }
                 }
             }
```

Comment:
That is actually not needed, I think. I have a comment here: https://github.com/dotnet/roslyn/pull/12927/files/55b4181c4040cae49c209fc23dd9b8e829cefb55#diff-8ab92731d875a419c8808cd50f634912R92

Let me know if I am wrong here.